blog/_posts/2018-03-29-missing.md

```diff
@@ -0,0 +1,410 @@
+   Base, in a package or in user code.
+
+2. Standard Julia code working with missing values is efficient, without special tricks.
+
+3. Missing values are safe by default: when passed to most functions, they either
+   propagate or throw an error.
+
+Finally, current limitations and future improvements are discussed.
+
+## A generic representation
+
+One of Julia's strengths is that user-defined types are as powerful and fast as built-in
+types. To fully take advantage of this, missing values had to support not only standard
+types like `Int`, `Float64` and `String`, but also any custom type. For this reason,
+Julia cannot use the so-called *sentinel* approach like R and Pandas to represent
+missingness, that is reserving special values within a type's domain. For example,
+R represents missing values in integer and boolean vectors using the smallest
+representable 32-bit integer (`-2,147,483,648`), and missing values in floating point
+vectors using a specific `NaN` payload (`1954`, which rumour says refers to Ross Ihaka's
+year of birth). Pandas only supports missing values in floating point vectors,
+and conflates them with `NaN` values.
+
+In order to provide a consistent representation of missing values which can be combined
+with any type, Julia 0.7 will use `missing`, an object with no fields which is the only
+instance of the the `Missing` singleton type. This is a normal Julia type with a few
+peculiarities which are detailed below. Values which can be either of type `T` or missing
+can simply be declared as `Union{Missing,T}`. For example, a vector holding either integers
+or missing values is of type `Array{Union{Missing,Int},1}`:
+
+    julia> [1, missing]
+    2-element Array{Union{Missing, Int64},1}:
+    1
+    missing
+
+An interesting property of this approach is that `Array{Union{Missing,T}}` behaves just
+like a normal `Array{T}` as soon as missing values have been replaced or skipped
+(see below).
+
+As can be seen in the example above, promotion rules are defined so that concatenating
+values of type `T` and missing values gives an array with element type `Union{Missing,T}`
+rather than `Any`:
+
+    julia> promote_type(Int, Missing)
+    Union{Missing, Int64}
+
+These promotion rules are essential for performance, as we will now see.
+
+## An efficient representation
+
+Another of Julia's strengths is that one does not need to use tricks such as vectorized
+calls to make code fast. In this spirit, working with missing values had to be
+efficient without requiring special treatment. While the `Union{Missing,T}` approach
+would have been very inefficient in previous Julia versions, the situation has
+dramatically changed thanks to two improvements implemented in the compiler in
+Julia 0.7.
+
+The first improvement involves optimizations for small `Union` types.
+When type inference detects that a variable can hold values of multiple types but
+that these types are in limited number (as is the case for `Union{Missing,T}`),
```

Comment:
This would be a good time to be clear on what this actually means so that way it can be referenced in the future. What is a small union type? What kinds of types can this be done with? Is it only unions of two bitstypes that this works with? I would like to see a footnote here.

Reply:
I'm not sure I'm the best person to document this, but I've added a note reflecting my (limited) understanding of how it works. I'd appreciate if others could confirm it's correct.

Reply:
Yeah, we'd need @vtjnash input on what the codegen limitations are here; I'm not sure if there's a limit on the # of union types that codegen will code-split on. Or maybe @Keno can comment on what the new compiler/optimizer does in union code-splitting?

Reply:
See the footnote I added, there's a constant for that. But it would be good to have somebody check that it's correct.